tests/test_json_to_parquet.py

```diff
@@ -0,0 +1,531 @@
+
+    @pytest.fixture()
+    def json_s3_objects(self, datadir, artifact_bucket, namespace):
+        s3_client = boto3.client("s3")
+        dataset_prefix = os.path.join(namespace, "tests")
+        object_keys = []
+        for dirpath, _, filenames in os.walk(datadir):
+            if len(filenames) > 0:
+                object_prefix = os.path.join(
+                        dataset_prefix,
+                        os.path.join(*dirpath.split("/")[-6:]))
+                for filename in filenames:
+                    object_key = os.path.join(object_prefix, filename)
+                    s3_client.upload_file(
+                            Filename=os.path.join(dirpath, filename),
+                            Bucket=artifact_bucket,
+                            Key=object_key)
+                    object_keys.append(object_key)
+        return object_keys
+
+    @pytest.fixture(scope="class")
+    def glue_context(self):
+        glue_context = GlueContext(SparkSession.builder.getOrCreate())
+        return glue_context
+
+    def test_upload_test_data(self, glue_test_data):
+        """
+        Upload test data to S3.
+        """
+        pass
+
+    def test_run_glue_crawler(self, glue_crawler):
+        """
+        Create and run Glue crawler.
+        """
+        pass
+
+    def test_get_table(self, glue_database_name, glue_flat_table_name,
+                       glue_nested_table_name, glue_context):
+        flat_table = json_to_parquet.get_table(
+                table_name=glue_flat_table_name,
+                database_name=glue_database_name,
+                glue_context=glue_context
+        )
+        assert flat_table.count() == 1
+        assert len(flat_table.schema().fields) == 2
+        nested_table = json_to_parquet.get_table(
+                table_name=glue_nested_table_name,
+                database_name=glue_database_name,
+                glue_context=glue_context
+        )
+        assert nested_table.count() == 1
+        assert len(nested_table.schema().fields) == 4
+
+    def test_has_nested_fields(self, glue_database_name, glue_flat_table_name,
+                               glue_nested_table_name, glue_context):
+            flat_table = json_to_parquet.get_table(
+                    table_name=glue_flat_table_name,
+                    database_name=glue_database_name,
+                    glue_context=glue_context
+            )
+            flat_table_schema = flat_table.schema()
+            assert not json_to_parquet.has_nested_fields(flat_table_schema)
+            nested_table = json_to_parquet.get_table(
+                    table_name=glue_nested_table_name,
+                    database_name=glue_database_name,
+                    glue_context=glue_context
+            )
+            nested_table_schema = nested_table.schema()
+            assert json_to_parquet.has_nested_fields(nested_table_schema)
+
+    def test_add_index_to_table(self, glue_database_name, glue_database_path,
```

Comment:
I see multiple asserts here for different components. Could separate into multiple tests and add into the name of the test function what it's testing for within `add_index_to_table`. Maybe some of these can be fixtures as I see it's dependent on each other.

Reply:
The `tables_with_index` dict which caches the results of each call to `json_to_parquet.add_index_to_table` is passed to subsequent calls to `json_to_parquet.add_index_to_table`. So these cases are more than dependent on the same input (like `nested_table_relationalized`), they are nested. I.e., they're dependent on the results of one another. It's still possible to split them up, but it takes the same code and splits it into different fixtures which makes it less readable imo.